### PR TITLE
Feature/docs sendfile

### PR DESCRIPTION
**Pull request by jordij**
I've tried implementing the feature described in #1115
Not sure if this is what you guys were after, I've added the sendfile module and tweak the serve view to use it. The setting SENDFILE_BACKEND can be set to development, simple or more advanced mods (xsendfile, wsgi etc..) up to the user to set the preferred one. I've updated the test suite with the method that tests the serve view.

### CODE REVIEW

**Comment by kaedroho** [comment]
Thanks @jordij. It looks like this is based on quite an old version of wagtail. Could you rebase it?

**Comment by jordij** [comment]
Oh boy, I think I did it on an old fork. Will update it and send it through

**Comment by zerolab** [comment]
Looks like you forgot to add sendfile to tox.ini

File: wagtail/wagtaildocs/tests.py [hunk 1]
```diff
@@ -142,6 +142,46 @@ def test_post(self):
...
+
+        self.client.logout()
+
+        # Serve document
+        self.document = Document.objects.get(title="Test document")
+
+        response = self.client.get(reverse('wagtaildocs_serve', args=(self.document.id, self.document.filename)))
+
+        self.assertEqual(response.status_code, 200)
+        self.assertEquals(
```
Comment (kaedroho) [inline wagtail/wagtaildocs/tests.py]: Could you change these to ``assertEqual``

**Comment by jordij** [comment]
I've merged the test method for the serve view as I didn't realise you guys had on in place already.
Cheers

File: wagtail/tests/settings.py [hunk 2]
```diff
@@ -136,5 +137,7 @@
 except ImportError:
     pass
 
+# Sendfile dev backend, do NOT use in production https://github.com/johnsensible/django-sendfile#django-sendfile
+SENDFILE_BACKEND = 'sendfile.backends.development'
```
Comment (kaedroho) [inline wagtail/tests/settings.py]: Looks like this is now a required setting: https://github.com/johnsensible/django-sendfile/blob/master/sendfile/__init__.py#L27-L29

**Comment by kaedroho** [comment]
Does django-sendfile support Python 3? They don't advertise support for it anywhere and don't have any automated testing setup showing support for it...

**Comment by jordij** [comment]
Hi guys,

I've been reading the threads on that sendfile repo and it's supposed to work even if it's not "officially" supported. I'll try contacting the maintainer, who appears to be too busy to merge some pull requests, and see if we can have an official release supporting Python 3.

If that doesn't work out let's do a fallback case with the old code you guys had?

Cheers

**Comment by kaedroho** [comment]
I added a few tests for the sendfile backends here: https://github.com/kaedroho/wagtail/commit/06c4d082eec7b46e10a6983085ac1da2fffb3e31

Could you pull this commit into this PR? This is so we can test sendfile against the versions of Python we support (as they don't currently have their own build matrix).

**Comment by kaedroho** [comment]
This works well on Python 3. One issue still remains though...

Currently, the docs serve view will crash if the developer hasn't specified a ``SENDFILE_BACKEND``, which I think could be an issue for sites being upgraded (I doubt many developers expect their document downloads to stop working after a Wagtail update...).

Ideally, Wagtail would fall back to using sendfiles "simple" backend if the developer hasn't set a ``SENDFILE_BACKEND`` as this is closest to the past behaviour. But this isn't easy to do as we can't currently specify a default backend for ``sendfile()``.

I've submitted a PR. I'm not sure if it will get accepted or released in a version soon though https://github.com/johnsensible/django-sendfile/pull/33

We could copy the ``sendfile()`` function into Wagtail utils and make the tweaks there. We could also copy the "simple" backend there so there's no need to download ``django-sendfile`` when it's not needed.

**Comment by kaedroho** [comment]
I've added some more changes to: https://github.com/kaedroho/wagtail/tree/jordij-feature/docs-sendfile

This pulls the ``sendfile()`` function into wagtail utils where we can make the backend tweak. I also noticed that sendfiles "simple" backend doesn't stream the file so I've also added a new backend to do that.

Let me know if you're happy with my changes.

**Comment by tomdyson** [comment]
@kaedroho I like this approach.

**Comment by davecranwell** [comment]
Hi @jordij, please see @kaedroho's message above. Are you happy with these changes? If so please merge his changes into your branch and comment here.

**Comment by MechanisM** [comment]
Going to check this out with uWSGI's sendfile - it can behave like apache's sendfile or like nginx's x-accel-redirect.

**Comment by jordij** [comment]
Yeah sure, I've merged the changes and updated the PR :)

**Comment by kaedroho** [comment]
Looks good to me :+1:

This should also solve #1183 as we can now specify a non-streaming backend when using static site gen.

Sendfile also sets the correct mime type for the file (#733) and uses the if-modified-since header.

**Comment by kaedroho** [comment]
Merged in c8d7f5baadfce3bfd9fd6ff0fc1b172ac030eec8. Thanks!

**Pull request closed without merging**